### PR TITLE
Data collator is working. Needs further testing but it should be fine

### PR DESCRIPTION
**Pull request by LouisCastricato**
Ported data collator to JAX. Working as expected when compared to https://github.com/huggingface/transformers/pull/12370 however I am not sure if this [the original] PR meets specification as it is a WIP.

### CODE REVIEW

File: data_collator.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        if self.tokenizer.mask_token is None:
+            raise ValueError
+
+    def __call__(self, examples: List[Union[List[int], jnp.ndarray, Dict[str, jnp.ndarray]]]
+                 ) -> Dict[str, jnp.ndarray]:
+        # Handle dict or lists with proper padding and conversion to tensor.
+        if isinstance(examples[0], (dict, BatchEncoding)):
+            batch = self.tokenizer.pad(examples, return_tensors="jax", pad_to_multiple_of=self.pad_to_multiple_of)
+        else:
+            batch = {"input_ids": _collate_batch(examples, self.tokenizer, pad_to_multiple_of=self.pad_to_multiple_of)}
```
Comment (morganmcg1) [inline data_collator.py]: We also need to return the original input for the `decoder_input_ids` for the decoder input, maybe rename like so:

```
batch = {"decoder_input_ids": _collate_batch(examples, self.tokenizer, pad_to_multiple_of=self.pad_to_multiple_of)}
```

And then pass `decoder_input_ids` to `self.mask_tokens` below

That way `batch` will include `input_ids`, `decoder_input_ids` and `labels`

File: data_collator.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            masked_indices[remaining, 1] += 1
+            for mi in masked_indices:
+                mask[tuple(mi)] = True
+            lengths -= 1
+            remaining = (lengths > 0) & (masked_indices[:, 1] < max_index)
+
+        # place the mask tokens
+        mask[np.where(special_tokens_mask==True)] = False
+        inputs_copy[np.where(mask==1)] = self.tokenizer.mask_token_id
+        labels[np.where(mask == 0)] = -100
```
Comment (morganmcg1) [inline data_collator.py:122]: Looking at the labels output, they're all -100, guessing something here needs to be fixed

```
[[-100 -100 -100 ... -100 -100 -100]
 [-100 -100 -100 ... -100 -100 -100]
 [-100 -100 -100 ... -100 -100 -100]]
```

**Review by morganmcg1 — CHANGES_REQUESTED** [review]

File: data_collator.py (hunk 1 above)
Comment (LouisCastricato) [reply data_collator.py]: Adding this now

File: data_collator.py (hunk 2 above)
Comment (LouisCastricato) [reply data_collator.py:122]: Resolved

File: data_collator.py (hunk 1 above)
Comment (LouisCastricato) [reply data_collator.py]: What do you mean by
> And then pass ```decoder_input_ids``` to ```self.mask_tokens below```

?

**Comment by morganmcg1** [comment]
I noticed that the infilling mask always seems to start with a space (e.g. "was<mask>" instead of "was <mask>"), I'll go ahead and merge this, but might be something we'd like to look at again.

Also, it looks like padding isn't done for the `decoder_input_ids` ?

```
lorem_ispum = """I ran down to the beach with the dog. It was a wet day, greay and fun. But sill, I had a good time"""

inputs = tokenizer(lorem_ispum, max_length=1024, padding=True, truncation=True)
collator = DataCollatorForTextInfilling(tokenizer)
out = collator(inputs)

print(tokenizer.decode(out['input_ids'][0]))
print(tokenizer.decode(out['decoder_input_ids'][0]))

>> <s>I ran<mask> the beach with<mask> It was a wet day, greay and fun. But sill, I had<mask> good time</s><pad><pad><pad>
>> <s>I ran down to the beach with the dog. It was a wet day, greay and fun. But sill, I had a good time</s>
```
